# Analyzing Hacker News Posts: Ask HN vs. Show HN


In this data analysis project, we will explore a dataset of Hacker News posts to compare the engagement levels between "Ask HN" and "Show HN" posts. "Ask HN" posts are used by users to ask the Hacker News community specific questions, while "Show HN" posts are used to showcase projects or interesting topics.

We will investigate the following questions:
Do "Ask HN" or "Show HN" posts receive more comments on average?
Is there a correlation between the time a post is created and the number of comments it receives?

By analyzing these questions, we aim to provide insights into the engagement dynamics of the Hacker News community and uncover trends that may help users optimize their posts for greater visibility and interaction.

In [1]:
from csv import reader
open_file=open('hacker_news.csv')
read_file=reader(open_file)
hn=list(read_file)

print(hn[0:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


## Removing the Headers from a list of lists

In [2]:
headers=hn[0]
print(headers)
hn=hn[1:]
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Extracting Ask HN & Show HN posts
By extracting the Ask HN & Show HN posts into differnt lists, we can analyse them easily.

In [3]:
ask_posts=[]
show_posts=[]
other_posts=[]

for row in hn:
    title=row[1]
    title=title.lower()
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

With the seperated posts of Ask HN & Show HN, we can calculate the average for them

In [4]:
total_ask_comments=0
for row in ask_posts:
    num_comments=int(row[4])
    total_ask_comments+=num_comments
    
avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [5]:
total_show_comments=0
for row in show_posts:
    num_comments=int(row[4])
    total_show_comments+=num_comments
avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


On an average Ask posts are receiving approximately 14 comments, whereas Show posts receive only 10 comments. SInce Ask posts are getting more comments let's do the remaining analysis on Ask posts.

## Finding the Amount of Ask Posts and Comments by Hour Created

Now let's determine if we can maximize our comments by posting at a certain time. First we will calculate the amount of ask posts created in each hour of the day, along with the number of comments received. Second Calculate the average number of comments ask posts receive by hour created.

In [6]:
import datetime as dt
result_list=[]
for row in ask_posts:
    result_list.append([row[6], int(row[4])])

result_list

counts_by_hour={}
comments_by_hour={}

for result in result_list:
    created=result[0]
    no_comment=result[1]
    created_time=dt.datetime.strptime(created, "%m/%d/%Y %H:%M")
    time=created_time.strftime("%H")
    if time in counts_by_hour:
        counts_by_hour[time]+=1
        comments_by_hour[time]+=no_comment
    else:
        counts_by_hour[time]=1
        comments_by_hour[time]=no_comment
    
comments_by_hour

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

## Calculating the Average number of comments for Ask HN Posts by Hour

With the two dictionaries we created, let's calculate the number of average comments of Ask HN Posts. 

In [7]:
avg_by_hour=[]

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

## Sorting and Printing Values a List of Lists

In [8]:
swap_avg_by_hour=[]

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)
sorted_swap=sorted(swap_avg_by_hour, reverse=True)

sorted_swap



[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [9]:
#let's print the 5 hours with the highest comments

print("Top 5 Hours for Ask Posts Comments")
for avg,hr in sorted_swap[:5]:
    print("{}: {:.2f} average comments per post".format(dt.datetime.strptime(hr,"%H").strftime("%H:%M"), avg))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour the receives the most comments per post is 15.00 with an average comments of 38.59. Betweeen the first and seconds hours there is 60% gap is there which proves that the 15.00 is the perfect time to make posts. 

The timezone used in the dataset is EST in the US. We could also use the time 15.00 as 3.00 PM EST.

## Conclusion:

Through our data analysis project exploring Hacker News posts, we have gained valuable insights into the engagement dynamics of the platform. By comparing "Ask HN" and "Show HN" posts, we discovered that "Ask HN" posts tend to receive more comments on average compared to "Show HN" posts. This suggests that users are more inclined to engage with posts that pose specific questions to the Hacker News community.

Furthermore, our analysis of the correlation between post creation time and comment count revealed an interesting trend. Posts created at 15:00 (3:00 PM) receive the highest average number of comments, indicating that this time may be optimal for maximizing post visibility and engagement. This finding suggests that timing plays a significant role in post success on Hacker News, and users may benefit from strategically timing their posts to coincide with peak engagement periods.